### DISCUSSION TITLE
Support Laravel Scout

### DISCUSSION THREAD

**Discussion by h-o-sein** [rappasoft/laravel-livewire-tables#1043] [Feature Requests]
Hi
Please Support Laravel Scout on livewire-table

**Comment by lrljoe** [comment]
You could emit an event containing a list of IDs, and use that in the Builder function.

Implementing Scout searching instead of the present method,  to my mind would somewhat defeat the point / cleverness of this package.

**Comment by booni3** [comment]
You can do this by extending the `executeQuery` method:

```php
public function executeQuery()
{
    if($searchTerm = $this->getSearch()){
        $model = $this->builder()->getModel();
        return $model::search($searchTerm, function(Indexes $index, $query, $options){
//                $options['sort'] = ["received_at:desc"];
            return $index->search($query, $options);
        })->simplePaginate($this->getPerPage() === -1 ? $this->getBuilder()->count() : $this->getPerPage(), $this->getComputedPageName());
    }

    parent::executeQuery();
}
```

**Reply by lrljoe** [reply]
At first glance,  it doesn't seem that this will work with this component's filters/sorting methods, I may be wrong tho!

If not, it kind of defeats the entire purpose behind this component, and you're better off just creating the table layout yourself and saving your memory/CPU.

**Reply by booni3** [reply]
You are correct. That specific example does not consider filter or sort.

Filter & sort is specific to the search engine you use. With Meilisearch, for example, you need to expose the index via a callback and add your filters here. Both filter and sort need to be constructed in a specific way.

There is a use case for scout implementation inside this package. In my case, I have around 10 tables in my app. Most of them are small and can use SQL with no problem. Two have a few million rows, and I need to search within full JSON documents. SQL cannot do this.

I want continuity between my tables, with the same look and feel. I also want to avoid remaking each component, such as the filtering system.

I have created a trait that I can drop into the required tables, and it will search with Scout without any additional configuration. You can also extract the filters and sorts from the query within this trait. If you remove it, you can use SQL again without any changes.

```php
trait WithScoutMeilisearch

    public function executeQuery()
    {
        if($searchTerm = $this->getSearch()){
            /** @var $model */
            $model = $this->builder()->getModel();

            /** @var Paginator $results */
            $results = $model::search($searchTerm, function(Indexes $index, $query, $options){
                // Make sorts from the query
                $options['sort'] = collect($this->getBuilder()->getQuery()->orders)->map(function(array $order){
                    return sprintf("%s:%s", $order['column'], $order['direction']);
                })->toArray();

                // Make filters from the query. i.e. `channel - "amazon" AND email = "test@example.com"`
                $options['filter'] = collect($this->getBuilder()->getQuery()->wheres)->map(function(array $filter){
                    return sprintf("%s = \"%s\"", $filter['column'], $filter['value']);
                })->implode(' AND ');

                // Execute search
                return $index->search($query, $options);

            })->simplePaginate($this->getPerPage() === -1 ? $this->getBuilder()->count() : $this->getPerPage(), $this->getComputedPageName());

            // Eager loading
            $results->load(array_keys($this->builder()->getEagerLoads()));

            return $results;
        }

        return parent::executeQuery();
    }
}
```

Within the data table component, you use the features such as filter as normal:

```php
public function filters(): array
{
    return [
        SelectFilter::make('Channel')
            ->options([
                '' => 'All',
                'amazon' => 'Amazon',
                'shopify' => 'Shopify',
            ])
            ->filter(function(Builder $builder, string $value) {
                $builder->where('channel', $value);
            })
    ];
}
```

There are other elements we could include in the trait, such as configuring searchable columns, but at the moment, I configure this all on the Meilisearch index directly. If you were using Agolia, you would likely do this best via their dashboard.

**Reply by lrljoe** [reply]
If you fancy writing that up in a quick set of guidance notes then I think they could fit in nicely with the existing docs for anyone else who's interested down the line.  Just create a new item in the docs folder and we can clean it up if necessary.

Or if you ensure the core functionality is available via the trait, then feel free to open a PR to introduce it into core and it'll be reviewed for inclusion.

I'll definitely chat with Anthony about options for smoother integrations for this in V3 too.

MySQL 8 does actually support json indexing, searching and dynamic columns by the way, it's a little awkward to implement when you first launch into it. MariaDB doesn't offer the same support as far as I know.
[MySQL 8 JSON Search](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html)
